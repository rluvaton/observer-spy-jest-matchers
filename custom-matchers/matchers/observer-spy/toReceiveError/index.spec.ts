import matcher from './index';
import { concat, EMPTY, of, throwError } from 'rxjs';

expect.extend(matcher);

// We aren't using a constants because if in some case the values changed (which it shouldn't)
// we don't want it to effect other tests.
const getObservableValue = (): any[][] => [
  // Start With Falsy
  [[0, 1, 2, 4, 5]],
  // Start with Truthy
  [[1, 2, 3, 4, 5]],

  // Start With Falsy
  [['', 'hello', 'this', 'is', 'so', 'cool']],
  // Start With Truthy
  [['hello', 'this', 'is', 'so', 'cool']],

  // Start with true
  [[true, false, true, false, true]],
  // Start with false
  [[false, true, true, false, false]],

  [[{}, { a: 'a' }, { b: 'b' }, { c: 'c' }]],
  [[{ a: 'a' }, {}, { b: 'b' }, { c: 'c' }]],

  [[[], ['I'], ['am'], ['an'], ['Array']]],
  [[['I'], [], ['am'], ['an'], ['Array']]],
];

describe('.toReceiveError', () => {
  describe('pass', () => {
    it('should passes when observable is only throwing error', async () => {
      // Arrange
      const obs = throwError(new Error('I dont care what the error is'));

      // Act & Assert
      await expect(obs).toReceiveError();
    });

    it('should passes when passing an error type that is the same as the error that throwen in the observable', async () => {
      // Arrange
      const obs = throwError(new Error('I dont care what the error is'));

      // Act & Assert
      await expect(obs).toReceiveError(Error);
    });

    describe('throw at the start', () => {
      it.each(getObservableValue())(
        'should passes when observable throw and then emit these values - %j',
        async (observableValues) => {
          // Arrange
          const obs = concat(throwError(new Error('Last error')), of(...observableValues));

          // Act & Assert
          await expect(obs).toReceiveError();
        },
      );
    });

    describe('throw in the middle', () => {
      it.each(getObservableValue())(
        'should passes when observable throw in the middle of emitting - %j',
        async (observableValues) => {
          // Arrange
          const firstSection = observableValues.slice(0, 2);
          const secondSection = observableValues.slice(2);

          const obs = concat(of(...firstSection), throwError(new Error('Last error')), of(...secondSection));

          // Act & Assert
          await expect(obs).toReceiveError();
        },
      );
    });

    describe('throw at the end', () => {
      it.each(getObservableValue())(
        'should passes when observable emit these values - %j and then throw in the end',
        async (observableValues) => {
          // Arrange
          const obs = concat(of(...observableValues), throwError(new Error('Last error')));

          // Act & Assert
          await expect(obs).toReceiveError();
        },
      );
    });
  });

  describe('fail', () => {
    it('should fail if a given observable is empty', async () => {
      // Arrange
      const obs = EMPTY;

      // Act & Assert
      await expect(() => expect(obs).toReceiveError()).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should fail when passing an error type that is not the same as the error that throwen in the observable', async () => {
      // Arrange
      class SubClass extends Error {}

      const obs = throwError(new Error('I dont care what the error is'));

      // Act & Assert
      await expect(() => expect(obs).toReceiveError(SubClass)).rejects.toThrowErrorMatchingSnapshot();
    });

    it.each(getObservableValue())(
      `should fail when an observable don't throw and only emit these values - %j`,
      async (observableValues) => {
        // Arrange
        const obs = of(...observableValues);

        // Act & Assert
        await expect(() => expect(obs).toReceiveError()).rejects.toThrowErrorMatchingSnapshot();
      },
    );
  });
});

describe('.not.toReceiveError', () => {
  describe('fail', () => {
    it('should fail when observable is only throwing error', async () => {
      // Arrange
      const obs = throwError(new Error('I dont care what the error is'));

      // Act & Assert
      await expect(() => expect(obs).not.toReceiveError()).rejects.toThrowErrorMatchingSnapshot();
    });

    describe('throw at the start', () => {
      it.each(getObservableValue())(
        'should fail when observable throw and then emit these values - %j',
        async (observableValues) => {
          // Arrange
          const obs = concat(throwError(new Error('Last error')), of(...observableValues));

          // Act & Assert
          await expect(() => expect(obs).not.toReceiveError()).rejects.toThrowErrorMatchingSnapshot();
        },
      );
    });

    describe('throw in the middle', () => {
      it.each(getObservableValue())(
        'should fail when observable throw in the middle of emitting - %j',
        async (observableValues) => {
          // Arrange
          const firstSection = observableValues.slice(0, 2);
          const secondSection = observableValues.slice(2);

          const obs = concat(of(...firstSection), throwError(new Error('Last error')), of(...secondSection));

          // Act & Assert
          await expect(() => expect(obs).not.toReceiveError()).rejects.toThrowErrorMatchingSnapshot();
        },
      );
    });

    describe('throw at the end', () => {
      it.each(getObservableValue())(
        'should fail when observable emit these values - %j and then throw in the end',
        async (observableValues) => {
          // Arrange
          const obs = concat(of(...observableValues), throwError(new Error('Last error')));

          // Act & Assert
          await expect(() => expect(obs).not.toReceiveError()).rejects.toThrowErrorMatchingSnapshot();
        },
      );
    });
  });

  describe('pass', () => {
    it('should pass if a given observable is empty', async () => {
      // Arrange
      const obs = EMPTY;

      // Act & Assert
      await expect(obs).not.toReceiveError();
    });

    it.each(getObservableValue())(
      `should pass when an observable don't throw and only emit these values - %j`,
      async (observableValues) => {
        // Arrange
        const obs = of(...observableValues);

        // Act & Assert
        await expect(obs).not.toReceiveError();
      },
    );
  });
});
