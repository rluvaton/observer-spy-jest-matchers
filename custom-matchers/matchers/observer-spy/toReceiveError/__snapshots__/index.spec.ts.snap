// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.not.toReceiveError fail should fail when observable is only throwing error 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the end should fail when observable emit these values - ["","hello","this","is","so","cool"] and then throw in the end 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the end should fail when observable emit these values - ["hello","this","is","so","cool"] and then throw in the end 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the end should fail when observable emit these values - [["I"],[],["am"],["an"],["Array"]] and then throw in the end 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the end should fail when observable emit these values - [[],["I"],["am"],["an"],["Array"]] and then throw in the end 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the end should fail when observable emit these values - [{"a":"a"},{},{"b":"b"},{"c":"c"}] and then throw in the end 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the end should fail when observable emit these values - [{},{"a":"a"},{"b":"b"},{"c":"c"}] and then throw in the end 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the end should fail when observable emit these values - [0,1,2,4,5] and then throw in the end 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the end should fail when observable emit these values - [1,2,3,4,5] and then throw in the end 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the end should fail when observable emit these values - [false,true,true,false,false] and then throw in the end 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the end should fail when observable emit these values - [true,false,true,false,true] and then throw in the end 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the start should fail when observable throw and then emit these values - ["","hello","this","is","so","cool"] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the start should fail when observable throw and then emit these values - ["hello","this","is","so","cool"] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the start should fail when observable throw and then emit these values - [["I"],[],["am"],["an"],["Array"]] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the start should fail when observable throw and then emit these values - [[],["I"],["am"],["an"],["Array"]] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the start should fail when observable throw and then emit these values - [{"a":"a"},{},{"b":"b"},{"c":"c"}] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the start should fail when observable throw and then emit these values - [{},{"a":"a"},{"b":"b"},{"c":"c"}] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the start should fail when observable throw and then emit these values - [0,1,2,4,5] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the start should fail when observable throw and then emit these values - [1,2,3,4,5] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the start should fail when observable throw and then emit these values - [false,true,true,false,false] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw at the start should fail when observable throw and then emit these values - [true,false,true,false,true] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw in the middle should fail when observable throw in the middle of emitting - ["","hello","this","is","so","cool"] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw in the middle should fail when observable throw in the middle of emitting - ["hello","this","is","so","cool"] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw in the middle should fail when observable throw in the middle of emitting - [["I"],[],["am"],["an"],["Array"]] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw in the middle should fail when observable throw in the middle of emitting - [[],["I"],["am"],["an"],["Array"]] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw in the middle should fail when observable throw in the middle of emitting - [{"a":"a"},{},{"b":"b"},{"c":"c"}] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw in the middle should fail when observable throw in the middle of emitting - [{},{"a":"a"},{"b":"b"},{"c":"c"}] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw in the middle should fail when observable throw in the middle of emitting - [0,1,2,4,5] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw in the middle should fail when observable throw in the middle of emitting - [1,2,3,4,5] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw in the middle should fail when observable throw in the middle of emitting - [false,true,true,false,false] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.not.toReceiveError fail throw in the middle should fail when observable throw in the middle of emitting - [true,false,true,false,true] 1`] = `
"expect(received).not.toReceiveError(expected)

Expected observable to not receive an error"
`;

exports[`.toReceiveError fail should fail if a given observable is empty 1`] = `
"expect(received).toReceiveError(expected)

Expected observable to receive an error"
`;

exports[`.toReceiveError fail should fail when an observable don't throw and only emit these values - ["","hello","this","is","so","cool"] 1`] = `
"expect(received).toReceiveError(expected)

Expected observable to receive an error"
`;

exports[`.toReceiveError fail should fail when an observable don't throw and only emit these values - ["hello","this","is","so","cool"] 1`] = `
"expect(received).toReceiveError(expected)

Expected observable to receive an error"
`;

exports[`.toReceiveError fail should fail when an observable don't throw and only emit these values - [["I"],[],["am"],["an"],["Array"]] 1`] = `
"expect(received).toReceiveError(expected)

Expected observable to receive an error"
`;

exports[`.toReceiveError fail should fail when an observable don't throw and only emit these values - [[],["I"],["am"],["an"],["Array"]] 1`] = `
"expect(received).toReceiveError(expected)

Expected observable to receive an error"
`;

exports[`.toReceiveError fail should fail when an observable don't throw and only emit these values - [{"a":"a"},{},{"b":"b"},{"c":"c"}] 1`] = `
"expect(received).toReceiveError(expected)

Expected observable to receive an error"
`;

exports[`.toReceiveError fail should fail when an observable don't throw and only emit these values - [{},{"a":"a"},{"b":"b"},{"c":"c"}] 1`] = `
"expect(received).toReceiveError(expected)

Expected observable to receive an error"
`;

exports[`.toReceiveError fail should fail when an observable don't throw and only emit these values - [0,1,2,4,5] 1`] = `
"expect(received).toReceiveError(expected)

Expected observable to receive an error"
`;

exports[`.toReceiveError fail should fail when an observable don't throw and only emit these values - [1,2,3,4,5] 1`] = `
"expect(received).toReceiveError(expected)

Expected observable to receive an error"
`;

exports[`.toReceiveError fail should fail when an observable don't throw and only emit these values - [false,true,true,false,false] 1`] = `
"expect(received).toReceiveError(expected)

Expected observable to receive an error"
`;

exports[`.toReceiveError fail should fail when an observable don't throw and only emit these values - [true,false,true,false,true] 1`] = `
"expect(received).toReceiveError(expected)

Expected observable to receive an error"
`;

exports[`.toReceiveError fail should fail when passing an error type that is not the same as the error that throwen in the observable 1`] = `
"expect(received).toReceiveError(expected)

Expected observable to receive an error"
`;
